### ISSUE TITLE
[ENH] Catching failure of forecast fit (e.g., arima zero division error), naive fallback

### ISSUE THREAD

**Issue by RikStarmans** [sktime/sktime#3838]
**Is your feature request related to a problem? Please describe.**
I am fitting a lot of time series with a column ensemble forecaster.
Each forecaster is now a simple autoarima model.
IF the fit fails, I want to be able to have a backup forecaster. Now, the entire pipeline fails.

Autoarima can simply fail and raise a ZeroDivisionError. I tried differencing etc, still this seems to be inherent to the method.
I tried the StatsForecast and the PMDarima implementation.
You are supposed to capture it.
Some of my signals seem to change over time and then the ZeroDivisionError is triggered.
These are my settings for StatsForecastAutoarima
```python
defaults = {
            "seasonal": False,
            # Seasonality, temperature is highly seasonal, does not add value
            # regular settings
            "d": None,  # is should look for ultimate difference term
            "max_p": 12,
            "max_q": 5,
            "stepwise": True,  
            "trend": True,  # considers trend models
            "n_jobs": None,
            "parallel": False,  # TRUE NOT IMPLEMENTED
            "n_fits": int(nfits),
            "trace": False,  # print status of fits
        }
```

**Describe the solution you'd like**
I rather have a bad estimate or a naive estimate then a failure.
Also I want to know for which column the fit failed.

**Describe alternatives you've considered**
@aiwalter 
I can think of a solution: convert your columns into panel data, then run a grid search on ForecastByLevel with multiple models, if ARIMA fails there would still be another one

In my case this costs too much compute.  I have fixed it as follows.
If the fit fails, it falls back to naive. 
Other options could be to handle this in ColumnEnsembleForecaster or BaseForecaster.

```python
class ARIMAMOD(BaseForecaster):
    _tags = {
        "scitype:y": "univariate",  # which y are fine? univariate/multivariate/both
        "ignores-exogeneous-X": False,  # does estimator ignore the exogeneous X?
        "handles-missing-data": False,  # can estimator handle missing data?
        "y_inner_mtype": "pd.Series",  # which types do _fit, _predict, assume for y?
        "X_inner_mtype": "pd.DataFrame",  # which types do _fit, _predict, assume for X?
        "requires-fh-in-fit": False,  # is forecasting horizon already required in fit?
        "X-y-must-have-same-index": False,  # can estimator handle different X/y index?
        "enforce_index_type": None,  # index type that needs to be enforced in X/y
        "capability:pred_int": True,  # does forecaster implement predict_quantiles?
        "python_dependencies": "statsforecast",
    }

    def __init__(
        self,
        start_p: int = 2,
        d: Optional[int] = None,
        start_q: int = 2,
        max_p: int = 5,
        max_d: int = 2,
        max_q: int = 5,
        start_P: int = 1,
        D: Optional[int] = None,
        start_Q: int = 1,
        max_P: int = 2,
        max_D: int = 1,
        max_Q: int = 2,
        max_order: int = 5,
        sp: int = 1,
        seasonal: bool = True,
        stationary: bool = False,
        information_criterion: str = "aicc",
        test: str = "kpss",
        seasonal_test: str = "seas",
        stepwise: bool = True,
        n_jobs: int = 2,
        trend: bool = True,
        method: Optional[str] = None,
        offset_test_args: Optional[str] = None,
        seasonal_test_args: Optional[Dict] = None,
        trace: bool = False,
        n_fits: int = 94,
        with_intercept: bool = True,
        approximation: Optional[bool] = None,
        truncate: Optional[bool] = None,
        blambda: Optional[float] = None,
        biasadj: bool = False,
        parallel: bool = False,
        strategy="last",
        window_length=None,
    ):
        self.start_p = start_p
        self.d = d
        self.start_q = start_q
        self.max_p = max_p
        self.max_d = max_d
        self.max_q = max_q
        self.start_P = start_P
        self.D = D
        self.start_Q = start_Q
        self.max_P = max_P
        self.max_D = max_D
        self.max_Q = max_Q
        self.max_order = max_order
        self.seasonal = seasonal
        self.stationary = stationary
        self.information_criterion = information_criterion
        self.test = test
        self.seasonal_test = seasonal_test
        self.stepwise = stepwise
        self.n_jobs = n_jobs
        self.trend = trend
        self.method = method
        self.offset_test_args = offset_test_args
        self.seasonal_test_args = seasonal_test_args
        self.trace = trace
        self.n_fits = n_fits
        self.with_intercept = with_intercept
        self.approximation = approximation
        self.truncate = truncate
        self.blambda = blambda
        self.biasadj = biasadj
        self.parallel = parallel
        self.strategy = strategy
        self.sp = sp
        self.window_length = window_length
        self.armamodel = StatsForecastAutoARIMA(
            start_p,
            d,
            start_q,
            max_p,
            max_d,
            max_q,
            start_P,
            D,
            start_Q,
            max_P,
            max_D,
            max_Q,
            max_order,
            sp,
            seasonal,
            stationary,
            information_criterion,
            test,
            seasonal_test,
            stepwise,
            n_jobs,
            trend,
            method,
            offset_test_args,
            seasonal_test_args,
            trace,
            n_fits,
            with_intercept,
            approximation,
            truncate,
            blambda,
            biasadj,
            parallel,
        )
        self.naivemodel = NaiveForecaster(strategy, window_length, sp)
        self.naive = False
        super().__init__()

    def _fit(self, y, X=None, fh=None):
        try:
            self.naive = False
            return self.armamodel.fit(fh=fh, X=X, y=y)
        except ZeroDivisionError:
            print("failed, using naive")
            self.naive = True
            return self.naivemodel.fit(fh=fh, X=X, y=y)

    def _predict(self, fh, X=None):
        if self.naive:
            return self.naivemodel.predict(fh=fh, X=X)
        else:
            return self.armamodel.predict(fh=fh, X=X)
```

**Additional context**
Tested the code and it seems to work.

**Comment by aiwalter** [comment]
what you did is change the actual model algo. I dont get whats the problem with using panel data, bcz that is already a soultion we have for this. `ColumnEnsembleForecaster` can not isolate the different columns as a separate fit, this can be done with `ForecastByLevel`.

**Comment by RikStarmans** [comment]
@aiwalter, I looked into [ForecastingGridSearch](https://www.sktime.org/en/stable/api_reference/auto_generated/sktime.forecasting.model_selection.ForecastingGridSearchCV.html), i.e. ForecastByLevel(ForecastingGridSearchCV(...))
and see a lot of advantages.
Still, I thought this needs to run for every forecaster. I don't know which one is failing and have 300 time series.
I only want to perform gridseach if it fails. This does not seem to be the case.

**Comment by aiwalter** [comment]
we have `error_score` param in the grid search classes. This is handling the exceptions in `fit` and turns them in to warnings by default. For standalone forecasters I think its better if they just raise an exception when failing.

**Comment by fkiraly** [comment]
@RikStarmans, I think a better solution would be a wrapper like `FallbackAtFitFailure`, which is a forecaster that has one forecaster was an alternative option (e.g., naive by default) that is used in case of failure.

**Comment by RikStarmans** [comment]
@fkiraly  exactly.. that is what I am looking for... I could modify my forecaster a bit

**Comment by aiwalter** [comment]
`FallbackAtFitFailure` sounds a bit bulky 😄  I can think of just `FitFailedForecaster` or `FallbackForecaster`

> I can't use *kwargs and * args --> runtime error / forbidden by sktime design principles

for sth like `FallbackAtFitFailure` you dont need kwargs

> I need to define the parameters on the object --> statements like self.d = d
> This can get tricky if they overlap.

No, the signature would be e.g. `FallbackAtFitFailure(forecaster=ARIMA(), fallback_forecaster=NaiveForecaster())`

> I need a way to get the tags of the forecasters I use.

pls see our extension templates or other compositions who to deal with tags

**Comment by fkiraly** [comment]
the precise signature would be:

`FallbackAtFitFailure(forecaster, fallback=None)`

I would inherit from `_DelegatedForecaster` and set the delegate depending on whether there is a failure in the inner `_fit`.

I.e., if you inherit from `_DelegatedForecaster`, all you need to implement in `_fit`. 

For x/y types, I would use `datatypes.ALL_TIME_SERIES_MTYPES`, in order to ensure conversions happen only inside.

**Comment by RikStarmans** [comment]
Thanks  @fkiraly simplified a lot.
I implemented as desired;
  * I think it is also nice to store the exception or maybe at a verbosive setting to print a warning
  * I see a lot of super(ClassName, self).__init__() in sktime,  I think these one can use super().__init__()

I will file a push request.

```python
class FallbackAtFitFailure(_DelegatedForecaster):
    """Fits with backup forecasts if primary forecaster fails.

    Forecasters such as StatsForecastAutoARIMA can fail when fitting.
    The failure is captured by the defined backup forecaster.

    Parameters
    ----------
    forecaster : target sktime forecaster
    fallback : backup sktime forecaster
    """

    _tags = {
        "scitype:y": "both",
        "ignores-exogeneous-X": False,
        "handles-missing-data": False,
        "requires-fh-in-fit": False,
        "y_inner_mtype": ALL_TIME_SERIES_MTYPES,
        "X_inner_mtype": ALL_TIME_SERIES_MTYPES,
        "fit_is_empty": False,
    }

    def __init__(self, forecaster, fallback=None):
        self.forecaster = forecaster
        self.fallback = fallback
        self.last_exception = None
        super().__init__()

    def _fit(self, y, X=None, fh=None):
        try:
            self.estimator_ = self.forecaster
            return super()._fit(fh=fh, X=X, y=y)
        except Exception as excep:
            self.last_exception = excep
            self.estimator_ = self.forecaster
            return super()._fit(fh=fh, X=X, y=y)
```

**Comment by fkiraly** [comment]
@RikStarmans, apologies, did not see this - we expect code like this usually as a draft PR, not in an issue.

I would put it into `sktime.forecasting.compose._fallback`.

Please also ensure that:

* linting runs, so when we trigger the remote CI it won't break at first step - speeds up the review and merge!
* you ran `check_estimator` on this

😄 

Also comment: fitted parameters should all end in an underscore, i.e., `last_exception_`, or it will not be seen by the default `get_fitted_params`

**Comment by fkiraly** [comment]
This is now superseded by @ninedigits' implementation of the `FallbackForecaster`. I just remembered this issue and will add an author credit to @RikStarmans.

Implemented in #5779
